### PR TITLE
Adding rotting to dismembered limbs and organs

### PR DESCRIPTION
**Pull request by Shdorsh**
## About The Pull Request

This PR mainly adds the rot that we know from corpses and gibs to **dismembered** limbs. Since surgical removal should be a procedure done carefully, it doesn't trigger that. The component qdeletes itself automatically once the limb has been reimplanted. Organs spilling out are also subject to rotting.
There's also been a small tweak for the corpse decomposition: In case the body isn't organic or has been husked, the component is also qdeleted, which would stop making it run even if it would just return before triggering for the rest of the game. Also, the game now has a miasma manager, which is basically a list of turfs and how much miasma to add on each. It makes it so that the tick intensive add_gas() proc isn't called each time, especially in cases where bodies, organs, limbs and gibs are piled up in one turf.

## Why It's Good For The Game

It expands on the yet insignificant miasma subsystem and seems logical: since corpses and gibs rot, why not dismembered limbs/organs? Also gives more insentive to clean up after a fight and do surgery right instead of just hammering with a chainsaw. Currently, the quantity of miasma is so insignificant that even with a round ending in full chaos, it barely has any impact.

## Changelog
:cl: Shdorsh
tweak: Made limbs and organs rot and make miasma when dismembered after 2 minutes, like corpses or gibs.
tweak: made the rot component for corpses a bit more efficient by removing it when it would just return without triggering anyway
tweak: Added a miasma processor. Made all the miasma add_gas() run off of it.
/:cl:

### CODE REVIEW

**Comment by Barhandar** [comment]
Did you know that miasma creates constant atmos churn (read: lag)?

**Comment by Shdorsh** [comment]
That's why I added some more qdels for the corpses, at least.

**Comment by Shdorsh** [comment]
If you want to I can try and create a subsystem that just calculates all the miasma that has to be put on a turf together first and then just does assert gas with the whole thing. That way we can save on corpses/gibs/organs that are all on the same turf. This will need some time tho. What do you say?

**Comment by Shdorsh** [comment]
First time I get this "Unmanaged global var use detected in code, please use the helpers." from Travis. Can anyone plz help me out about what that means?

**Comment by habader** [comment]
Butchering?
Also 2 minutes seems a little short

**Comment by Shdorsh** [comment]
Whelp, it compiles normally on desktop but it seems travis doesn't like global vars that aren't managed. Problem is I need to change a lot of the stuff I coded now. Should I try again, cleaner?

File: code/datums/components/rot.dm [hunk 1]
```diff
@@ -14,14 +35,9 @@
...
-	if(!istype(T))
-		return
 
-	var/datum/gas_mixture/stank = new
-	ADD_GAS(/datum/gas/miasma, stank.gases)
-	stank.gases[/datum/gas/miasma][MOLES] = amount
-	T.assume_air(stank)
-	T.air_update_turf()
+	if(istype(T) && !istype(T,/turf/open/space))
+		GLOB.MIASMA_PROCESSOR.tilestodiffuse[T] += amount
```
Comment (ike709) [inline code/datums/components/rot.dm]: This is causing runtimes.
```
[16:22:59] Runtime in rot.dm,40: Cannot read null.tilestodiffuse
  proc name: process (/datum/component/rot/process)
  src: /datum/component/rot/gibs (/datum/component/rot/gibs)
  call stack:
  /datum/component/rot/gibs (/datum/component/rot/gibs): process(10)
  Processing (/datum/controller/subsystem/processing): fire(0)
  Processing (/datum/controller/subsystem/processing): ignite(0)
  Master (/datum/controller/master): RunQueue()
  Master (/datum/controller/master): Loop()
  Master (/datum/controller/master): StartProcessing(0)
```

Comment (Shdorsh) [reply code/datums/components/rot.dm]: so basically, it isn't fully initialized and glob_datum doesn't do that. how do I go about to initialize it?

**Comment by Shdorsh** [comment]
There we go, wew. Tested on homeserver, does its job.

**Comment by anconfuzedrock** [comment]
If this doesn't halt the rot with fridges and freezers, it really really should.

**Comment by Shdorsh** [comment]
Brb, good point

**Comment by Shdorsh** [comment]
There we go.

**Comment by Grazyn** [comment]
Do they rot into skeleton limbs? If not, they should

**Comment by Shdorsh** [comment]
for now, it only creates miasma. i'll keep the skelelimbs maybe for a later update

**Comment by davethwave** [comment]
Might be interesting to give the limbs clone damage after rotting for a while.

**Comment by Shdorsh** [comment]
Ok, will do the skeletization and clone damage later on. Just gimme some time.

**Comment by Barhandar** [comment]
Adding damage to limbs is how you get murder-by-adding-bodyparts and genetics just making monkeymen for disassembly (because it's still faster than getting someone to build organ grower).

Also, does organ harvester correctly not add the component? It's "surgical".

**Comment by RaveRadbury** [comment]
> Adding damage to limbs is how you get murder-by-adding-bodyparts

Sure, but if I attached a bunch of rotting limbs to you that _would probably kill you_ irl

**Comment by ExcessiveUseOfCobblestone** [comment]
why do you webbit if you're going to test locally anyways yikes

File: code/__DEFINES/atmospherics.dm [hunk 2]
```diff
@@ -304,3 +304,4 @@ GLOBAL_LIST_INIT(pipe_paint_colors, list(
 #define MIASMA_CORPSE_MOLES 0.02
 #define MIASMA_GIBS_MOLES 0.005
 #define MIASMA_HYGIENE_MOLES 0.002
+GLOBAL_DATUM_INIT(MIASMA_PROCESSOR, /datum/controller/miasma_processor, new())
```
Comment (ninjanomnom) [inline code/__DEFINES/atmospherics.dm]: What the fuck is this? Make an actual subsystem

File: code/datums/components/rot.dm [hunk 3]
```diff
@@ -1,3 +1,33 @@
+//This is supposed to help make miasma at certain turfs by reducing the number of atmos changes to do
+//Needed, since it would attach the processor to null, which would make the process() proc not trigger
+datum/controller/miasma_processor/New()
```
Comment (ninjanomnom) [inline code/datums/components/rot.dm]:
```suggestion
/datum/controller/miasma_processor/New()
```

File: code/datums/components/rot.dm [hunk 4]
```diff
@@ -46,14 +72,34 @@
...
 
 /datum/component/rot/gibs
 	amount = MIASMA_GIBS_MOLES
+
+/datum/component/rot/bodypart
+	amount = MIASMA_GIBS_MOLES
+
+/datum/component/rot/bodypart/process()
+	var/obj/item/bodypart/BP = parent
+	if(BP.owner || BP.burn_dam > 100)
```
Comment (ninjanomnom) [inline code/datums/components/rot.dm]: Instead of checking in process make it get removed in the process of being reattached/burning up/whatever event makes this get removed. Using process to check for changes in state is a bad idea

File: code/datums/components/rot.dm [hunk 5]
```diff
@@ -46,14 +72,34 @@
...
+		qdel(src)
+		return
+	..()
+
+/datum/component/rot/organ
+	amount = MIASMA_GIBS_MOLES
+
+/datum/component/rot/organ/process()
+	var/obj/item/organ/O = parent
+	if(O.owner)
```
Comment (ninjanomnom) [inline code/datums/components/rot.dm]: Same as above

File: code/modules/mob/living/carbon/death.dm [hunk 6]
```diff
@@ -34,8 +34,7 @@
 				if(no_brain || !istype(X, /obj/item/organ/brain))
 					qdel(X)
 		else //we're going to drop all bodyparts except chest, so the only organs that needs spilling are those inside it.
-			for(var/X in internal_organs)
-				var/obj/item/organ/O = X
+			for(var/obj/item/organ/O in internal_organs)
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/death.dm]: Why was this changed?

File: code/modules/mob/living/carbon/death.dm [hunk 7]
```diff
@@ -44,9 +43,9 @@
 					O.Remove(src)
 					O.forceMove(Tsec)
 					O.throw_at(get_edge_target_turf(src,pick(GLOB.alldirs)),rand(1,3),5)
+					O.start_rotting()
 	else
-		for(var/X in internal_organs)
-			var/obj/item/organ/I = X
+		for(var/obj/item/organ/I in internal_organs)
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/death.dm]: Why was this changed?

File: code/modules/mob/living/carbon/death.dm [hunk 8]
```diff
@@ -56,10 +55,10 @@
 			I.Remove(src)
 			I.forceMove(Tsec)
 			I.throw_at(get_edge_target_turf(src,pick(GLOB.alldirs)),rand(1,3),5)
+			I.start_rotting()
 
 
 /mob/living/carbon/spread_bodyparts()
-	for(var/X in bodyparts)
-		var/obj/item/bodypart/BP = X
+	for(var/obj/item/bodypart/BP in bodyparts)
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/death.dm]: Why was this changed?

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 9]
```diff
@@ -530,6 +530,15 @@
 		if(L)
 			L.update_icon()
 
+//Called by dismember(), meaning the organ wasn't carefully (surgically) removed, rather violently
+/obj/item/bodypart/proc/start_rotting()
+	if(status == BODYPART_ORGANIC)
+		addtimer(CALLBACK(src, .proc/rot_now), 2 MINUTES)
+
+/obj/item/bodypart/proc/rot_now()
+	desc += " It smells terrible."
```
Comment (ninjanomnom) [inline code/modules/surgery/bodyparts/bodyparts.dm]: Make the component listen for examine and add this itself

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 10]
```diff
@@ -530,6 +530,15 @@
 		if(L)
 			L.update_icon()
 
+//Called by dismember(), meaning the organ wasn't carefully (surgically) removed, rather violently
+/obj/item/bodypart/proc/start_rotting()
+	if(status == BODYPART_ORGANIC)
+		addtimer(CALLBACK(src, .proc/rot_now), 2 MINUTES)
```
Comment (ninjanomnom) [inline code/modules/surgery/bodyparts/bodyparts.dm]: This still happens even if the component should no longer be added

File: code/modules/surgery/bodyparts/dismemberment.dm [hunk 11]
```diff
@@ -22,6 +22,9 @@
 	SEND_SIGNAL(C, COMSIG_ADD_MOOD_EVENT, "dismembered", /datum/mood_event/dismembered)
 	drop_limb()
 
+	if(status == BODYPART_ORGANIC)
+		addtimer(CALLBACK(src, .proc/start_rotting), 2 MINUTES)
```
Comment (ninjanomnom) [inline code/modules/surgery/bodyparts/dismemberment.dm]: Same as above

File: code/modules/surgery/bodyparts/dismemberment.dm [hunk 12]
```diff
@@ -56,15 +59,15 @@
 	var/turf/T = get_turf(C)
 	C.add_splatter_floor(T)
 	playsound(get_turf(C), 'sound/misc/splort.ogg', 80, 1)
-	for(var/X in C.internal_organs)
-		var/obj/item/organ/O = X
+	for(var/obj/item/organ/O in C.internal_organs)
```
Comment (ninjanomnom) [inline code/modules/surgery/bodyparts/dismemberment.dm]: ?

File: code/modules/surgery/organs/organ_internal.dm [hunk 13]
```diff
@@ -98,6 +98,14 @@
 	else
 		..()
 
+/obj/item/organ/proc/start_rotting()
+	if(!synthetic)
+		addtimer(CALLBACK(src, .proc/rot_now), 2 MINUTES)
```
Comment (ninjanomnom) [inline code/modules/surgery/organs/organ_internal.dm]: Same thing

File: code/modules/surgery/organs/organ_internal.dm [hunk 14]
```diff
@@ -98,6 +98,14 @@
...
+
+/obj/item/organ/proc/rot_now()
+		desc += " It smells terrible."
```
Comment (ninjanomnom) [inline code/modules/surgery/organs/organ_internal.dm]: Same thing

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by optimumtact** [comment]
failed to address review comments
